File: code/backend/src/llm_geometry/lex/spectrum.py
"""Embedding geometry for the Lexicon Lab (FR-620..FR-623).

This module implements *exactly* the six steps of the "Spectrum" section of
`specs/006-lexicon-lab-tiny/architecture.md`, which is the contract the TypeScript
browser engine also implements (SC-605 holds the two to <= 1e-5):

1. **Column-mean-centre** ``Ac = A - mean(A, axis=0)``. Centring is why the maximum
   attainable rank is ``min(V-1, d)``: subtracting the column mean forces every row of
   ``Ac`` into the hyperplane orthogonal to the all-ones vector.
2. **Gram** ``G = Ac.T @ Ac``, shape ``(d, d)``, symmetric eigendecomposition. The
   eigenvalues of ``G`` *are* the squared singular values of ``Ac``, so there is no SVD
   anywhere in this file. Two reasons, both real: a ``d x d`` symmetric eigensolve is
   ~2 ms in a browser on the shapes this tab uses, and the source project crashes on
   this exact computation because ``torch.linalg.svdvals`` has no MPS kernel.
3. **Clamp** ``lambda_i <- max(lambda_i, 0)`` (negatives are float error only), sort
   descending, ``sigma_i = sqrt(lambda_i)``.
4. With ``p_i = lambda_i / sum(lambda)``: effective rank ``exp(-sum p ln p)``, stable
   rank ``sum(lambda)/lambda_1``, participation ratio ``1/sum(p^2)``, ``frac_var_top2``,
   ``frac_var_top10``, ``n_dims_for_90pct``.
5. The **ceiling** ``min(V-1, d)`` is reported alongside, never implied.
6. **PCA coordinates** ``Ac @ E[:, :3]`` for the top-3 eigenvectors of ``G``, with each
   component's explained-variance ratio. These are a *projection* and are labelled as
   such wherever they are drawn — unlike the Geometry Lab's sphere, which is native 3-D.

`random_baseline_spectrum` is the reason FR-622 exists: effective rank rises with ``|V|``
*for random matrices too*, purely because a taller matrix has more independent rows to
fill the ``min(V-1, d)`` ceiling with. Drawing an untrained model's spectrum next to a
trained one makes that mechanical confound visible instead of letting it masquerade as
learning (SC-604).

**Sign convention (an addition, documented because architecture.md is silent).**
Eigenvector signs are arbitrary — ``numpy.linalg.eigh`` and a browser eigensolver may
disagree on them, which would break the SC-605 golden test on `pca_coords` while every
statistic matched. We therefore fix each eigenvector's sign so its largest-magnitude
entry is positive (first index wins a tie). This changes nothing geometric: it is a
reflection of a coordinate axis.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Sequence

import numpy as np

from ..errors import ComputeError, InvalidParamError
from .config import DEFAULT_SEED, PCA_COMPONENTS

__all__ = [
    "Spectrum",
    "spectrum",
    "random_baseline_spectrum",
    "untrained_matrix",
    "model_matrix",
    "spectra_for_model",
    "compare_to_baseline",
    "effective_rank_of",
]


@dataclass(frozen=True)
class Spectrum:
    """Every statistic FR-620/FR-621/FR-622/FR-623 asks for, measured from one matrix.

    ``eigenvalues``/``singular_values``/``explained_variance`` are length ``d`` and sorted
    descending; entries beyond ``max_rank`` are exactly ``0.0`` up to float error.
    """

    rows: int
    """``V`` — the number of embedding rows the matrix has."""

    d_model: int
    """``d`` — the number of columns."""

    max_rank: int
    """``min(V-1, d)``: the ceiling centring imposes. FR-622 draws this."""

    eigenvalues: tuple[float, ...]
    singular_values: tuple[float, ...]
    explained_variance: tuple[float, ...]
    """``p_i = lambda_i / sum(lambda)``, so this sums to 1 (or is all-zero)."""

    total_variance: float
    """``sum(lambda_i)``, which is ``||Ac||_F^2``."""

    effective_rank: float
    stable_rank: float
    participation_ratio: float
    frac_var_top2: float
    frac_var_top10: float
    n_dims_for_90pct: int

    pca_coords: tuple[tuple[float, ...], ...]
    """``(V, 3)`` projection onto the leading eigenvectors. A PROJECTION (FR-623)."""

    pca_explained_variance_ratio: tuple[float, ...]

    degenerate: bool
    """True when the centred matrix is exactly zero (``V == 1``, or identical rows).

    Every ratio is then undefined, so the rank statistics are reported as ``0.0`` and
    this flag says why rather than letting a caller read ``0.0`` as a measurement.
    """

    def as_dict(self) -> dict[str, Any]:
        """Plain-JSON form; the routes round it to 6 significant digits."""
        return {
            "rows": self.rows,
            "d_model": self.d_model,
            "max_rank": self.max_rank,
            "eigenvalues": list(self.eigenvalues),
            "singular_values": list(self.singular_values),
            "explained_variance": list(self.explained_variance),
            "total_variance": self.total_variance,
            "effective_rank": self.effective_rank,
            "stable_rank": self.stable_rank,
            "participation_ratio": self.participation_ratio,
            "frac_var_top2": self.frac_var_top2,
            "frac_var_top10": self.frac_var_top10,
            "n_dims_for_90pct": self.n_dims_for_90pct,
            "pca_coords": [list(row) for row in self.pca_coords],
            "pca_explained_variance_ratio": list(self.pca_explained_variance_ratio),
            "degenerate": self.degenerate,
        }

    def summary(self) -> dict[str, Any]:
        """The scalar statistics only — what a baseline comparison needs."""
        return {
            "rows": self.rows,
            "d_model": self.d_model,
            "max_rank": self.max_rank,
            "effective_rank": self.effective_rank,
            "stable_rank": self.stable_rank,
            "participation_ratio": self.participation_ratio,
            "frac_var_top2": self.frac_var_top2,
            "frac_var_top10": self.frac_var_top10,
            "n_dims_for_90pct": self.n_dims_for_90pct,
            "total_variance": self.total_variance,
            "degenerate": self.degenerate,
        }


def spectrum(matrix: Any, *, n_components: int = PCA_COMPONENTS) -> Spectrum:
    """The spectrum of a ``(V, d)`` matrix, by the six steps above.

    ``matrix`` is anything ``numpy.asarray`` accepts with a 2-D float shape — a numpy
    array, a nested list, or a detached torch tensor. Computation is in float64
    regardless of the input dtype: the Gram matrix squares the condition number, and
    float32 there would cost about half the significant digits of every small eigenvalue.
    """
    a = np.asarray(matrix, dtype=np.float64)
    if a.ndim != 2:
        raise InvalidParamError(f"spectrum needs a 2-D (V, d) matrix, got shape {a.shape}")
    if a.size == 0:
        raise InvalidParamError(f"spectrum needs a non-empty matrix, got shape {a.shape}")
    if not np.all(np.isfinite(a)):
        raise ComputeError(
            "matrix contains non-finite values; refusing to report a spectrum of NaN/inf"
        )

    rows, d_model = a.shape
    max_rank = min(rows - 1, d_model)

    # 1. column-mean-centre
    centred = a - a.mean(axis=0, keepdims=True)

    # 2. d x d Gram + symmetric eigendecomposition (NO SVD)
    gram = centred.T @ centred
    # eigh returns ascending eigenvalues; reverse for descending.
    eigenvalues_asc, eigenvectors_asc = np.linalg.eigh(gram)
    eigenvalues = eigenvalues_asc[::-1].copy()
    eigenvectors = eigenvectors_asc[:, ::-1].copy()

    # 3. clamp float-error negatives, then sqrt
    np.maximum(eigenvalues, 0.0, out=eigenvalues)
    singular_values = np.sqrt(eigenvalues)

    total = float(eigenvalues.sum())
    n_keep = min(n_components, d_model)
    coords = _pca_coords(centred, eigenvectors, n_keep, n_components)

    if total <= 0.0:
        # Exactly-zero centred matrix: p_i is 0/0. Report zeros and flag it (see
        # `degenerate`) rather than inventing a rank for a matrix that has none.
        zeros = tuple(0.0 for _ in range(d_model))
        return Spectrum(
            rows=rows,
            d_model=d_model,
            max_rank=max_rank,
            eigenvalues=zeros,
            singular_values=zeros,
            explained_variance=zeros,
            total_variance=0.0,
            effective_rank=0.0,
            stable_rank=0.0,
            participation_ratio=0.0,
            frac_var_top2=0.0,
            frac_var_top10=0.0,
            n_dims_for_90pct=0,
            pca_coords=coords,
            pca_explained_variance_ratio=tuple(0.0 for _ in range(n_components)),
            degenerate=True,
        )

    # 4. the statistics
    p = eigenvalues / total
    positive = p[p > 0.0]
    effective_rank = float(np.exp(-np.sum(positive * np.log(positive))))
    stable_rank = float(total / eigenvalues[0])
    participation_ratio = float(1.0 / np.sum(p * p))
    cumulative = np.cumsum(p)
    n_dims_for_90pct = int(np.searchsorted(cumulative, 0.9) + 1)
    n_dims_for_90pct = min(n_dims_for_90pct, d_model)

    ratios = [float(p[i]) if i < d_model else 0.0 for i in range(n_components)]

    return Spectrum(
        rows=rows,
        d_model=d_model,
        max_rank=max_rank,
        eigenvalues=tuple(float(v) for v in eigenvalues),
        singular_values=tuple(float(v) for v in singular_values),
        explained_variance=tuple(float(v) for v in p),
        total_variance=total,
        effective_rank=effective_rank,
        stable_rank=stable_rank,
        participation_ratio=participation_ratio,
        frac_var_top2=float(p[:2].sum()),
        frac_var_top10=float(p[:10].sum()),
        n_dims_for_90pct=n_dims_for_90pct,
        pca_coords=coords,
        pca_explained_variance_ratio=tuple(ratios),
        degenerate=False,
    )


def _pca_coords(
    centred: np.ndarray,
    eigenvectors: np.ndarray,
    n_keep: int,
    n_components: int,
) -> tuple[tuple[float, ...], ...]:
    """``Ac @ E[:, :k]`` with the documented sign convention, right-padded with zeros.

    Padding only happens if ``d < n_components``, which the shipped ``D_MODEL_CHOICES``
    never trigger; it exists so the response shape is ``(V, n_components)`` unconditionally.
    """
    rows = centred.shape[0]
    basis = eigenvectors[:, :n_keep]
    if n_keep:
        # Sign convention: largest-magnitude entry of each eigenvector is positive.
        pivot = np.argmax(np.abs(basis), axis=0)
        signs = np.sign(basis[pivot, np.arange(n_keep)])
        signs[signs == 0.0] = 1.0
        basis = basis * signs
    projected = centred @ basis if n_keep else np.zeros((rows, 0))
    if n_keep < n_components:
        projected = np.hstack([projected, np.zeros((rows, n_components - n_keep))])
    return tuple(tuple(float(v) for v in row) for row in projected)


# -- the untrained baseline (FR-622 / SC-604) ---------------------------------------------


def untrained_matrix(
    config: Any, seed: int = DEFAULT_SEED, *, which: str = "embedding"
) -> np.ndarray:
    """The ``(V, d)`` matrix of a **freshly initialized, untrained** `LexModel`.

    The baseline must be the *same* initialization the trained model started from, so
    this builds a real model (``llm_geometry.lex.model.LexModel(config, seed=seed)``)
    rather than re-deriving ``N(0, 0.02²)`` here. A second copy of the init would be a
    second thing to keep in sync, and the first time the two drifted the baseline would
    quietly stop being a baseline.

    ``which`` is ``"embedding"`` (``model.embed``) or ``"readout"`` (``model.head_w``).
    A **tied** model has no separate readout and raises rather than returning the
    embedding under a second name — the source project's `--tie` bug was exactly that,
    logging one matrix as two spectra and inviting a comparison of a thing with itself.
    """
    from .model import LexModel  # local import: torch is not needed for the maths above

    return model_matrix(LexModel(config, seed=seed), which)


def model_matrix(model: Any, which: str = "embedding") -> np.ndarray:
    """Pull the ``(V, d)`` embedding or readout out of a `LexModel`, as float64.

    ``which="readout"`` on a **tied** model raises: a tied model has exactly one matrix
    and reporting it twice would dress a thing up as a comparison with itself.
    """
    if which == "embedding":
        param = model.embed
    elif which == "readout":
        if getattr(model.cfg, "tied", False):
            raise InvalidParamError(
                "this model is tied: its readout IS its embedding, so it has exactly one "
                "spectrum. Reporting it twice would dress one matrix up as a comparison."
            )
        param = model.head_w
    else:
        raise InvalidParamError(f'which must be "embedding" or "readout", got {which!r}')
    array = np.asarray(param.detach().cpu().numpy(), dtype=np.float64)
    if array.ndim != 2:
        raise ComputeError(f"{which} parameter has shape {array.shape}, expected (V, d)")
    return array


def random_baseline_spectrum(
    config: Any,
    seed: int = DEFAULT_SEED,
    *,
    which: str = "embedding",
    n_components: int = PCA_COMPONENTS,
) -> Spectrum:
    """The spectrum of an **untrained** model at the same shape (FR-622, SC-604).

    A trained model's effective rank means nothing on its own: a random ``(V, d)``
    Gaussian matrix already climbs toward ``min(V-1, d)`` as ``V`` grows, so a rank
    "staircase" against the vocabulary budget is *expected without any learning*. This
    is the control the panel draws behind the trained curve.
    """
    return spectrum(untrained_matrix(config, seed, which=which), n_components=n_components)


def compare_to_baseline(trained: Spectrum, baseline: Spectrum) -> dict[str, Any]:
    """The trained-vs-untrained deltas the spectrum panel states in words.

    ``effective_rank_delta`` is the honest headline number: trained minus random at the
    same shape. It can be negative — training often *concentrates* variance — and the
    panel says so rather than only celebrating increases.
    """
    if (trained.rows, trained.d_model) != (baseline.rows, baseline.d_model):
        raise InvalidParamError(
            f"baseline shape {(baseline.rows, baseline.d_model)} does not match "
            f"trained shape {(trained.rows, trained.d_model)}; a baseline at a "
            "different shape would not be a control"
        )
    return {
        "effective_rank_delta": trained.effective_rank - baseline.effective_rank,
        "stable_rank_delta": trained.stable_rank - baseline.stable_rank,
        "participation_ratio_delta": trained.participation_ratio - baseline.participation_ratio,
        "frac_var_top2_delta": trained.frac_var_top2 - baseline.frac_var_top2,
        "max_rank": trained.max_rank,
        "effective_rank_frac_of_ceiling": (
            trained.effective_rank / trained.max_rank if trained.max_rank else 0.0
        ),
        "baseline_effective_rank_frac_of_ceiling": (
            baseline.effective_rank / baseline.max_rank if baseline.max_rank else 0.0
        ),
    }


def spectra_for_model(
    model: Any, *, tied: bool, n_components: int = PCA_COMPONENTS
) -> dict[str, Spectrum]:
    """Every spectrum a model has: ``{"embedding": ...}``, plus ``"readout"`` when untied.

    A **tied** model reports exactly one spectrum, labelled tied. The source project logs
    E and U as two spectra for a tied model, which is the same matrix twice dressed up as
    a comparison; not repeating that is a stated requirement of this feature.
    """
    out = {"embedding": spectrum(model_matrix(model, "embedding"), n_components=n_components)}
    if not tied:
        out["readout"] = spectrum(model_matrix(model, "readout"), n_components=n_components)
    return out


def effective_rank_of(eigenvalues: Sequence[float]) -> float:
    """``exp(-sum p ln p)`` for an already-computed eigenvalue list (golden-test helper)."""
    lam = np.maximum(np.asarray(eigenvalues, dtype=np.float64), 0.0)
    total = float(lam.sum())
    if total <= 0.0:
        return 0.0
    p = lam[lam > 0.0] / total
    return float(np.exp(-np.sum(p * np.log(p))))

File: code/frontend/src/lib/lexEngine/spectrum.ts
/**
 * Embedding geometry: the six steps of `architecture.md` § Spectrum, in order.
 *
 * For a matrix `A` of shape (V, d) — the embedding, or the readout when untied:
 *
 *   1. column-mean-centre:  Ac = A - mean(A, axis=0)
 *   2. Gram:                G  = Ac^T Ac, shape (d, d); a SYMMETRIC eigendecomposition
 *                           gives lambda_i = sigma_i^2 directly
 *   3. clamp, sort descending, sigma_i = sqrt(lambda_i)
 *   4. with p_i = lambda_i / sum(lambda):
 *        effective_rank = exp(-sum p_i ln p_i)      (p_i = 0 contributes 0)
 *        stable_rank    = sum(lambda) / lambda_1
 *        participation  = 1 / sum(p_i^2)
 *        frac_var_top2, frac_var_top10, n_dims_for_90pct
 *   5. report the ceiling min(V-1, d) alongside — centring is why the ALGEBRAIC rank is
 *      bounded by V-1, and that bound is climbed even by random matrices (FR-622), so it
 *      is drawn rather than left implicit
 *   6. PCA coordinates Ac @ E[:, :3] with the explained-variance ratio of each component,
 *      eigenvector signs canonicalized. These are a PROJECTION and must be labelled as
 *      one (FR-623).
 *
 * **Effective rank is an entropy, not a count.** `min(|V|-1, d)` bounds it, but it
 * reaches that bound only when the spectrum is perfectly flat. Measured on untrained
 * models at d=128, the ALGEBRAIC rank hits the 128 ceiling exactly from the `first`
 * budget onward while the effective rank is still only ~104 at |V| = 314 and climbing
 * with decelerating increments. It does not plateau at the ceiling, and nothing in the
 * UI may say it does.
 *
 * **No SVD.** The d x d Gram route is what makes this ~2 ms in a browser, and it is also
 * what sidesteps the source project's crash (`torch.linalg.svdvals` has no MPS kernel).
 * d <= 128 here, so a cyclic Jacobi rotation sweep is both fast and simple, and unlike a
 * QR-iteration it needs no tridiagonalisation to get eigenvectors out.
 */

import { invalidParam } from "../geoEngine/errors";
import { sfc32 } from "./model";

/** Components kept for the PCA token cloud. */
export const PCA_COMPONENTS = 3;
/** Bars drawn in the spectrum plot (the leading singular values). */
export const SPECTRUM_DISPLAY_K = 48;

export interface Eigen {
  /** Eigenvalues, descending. */
  values: Float64Array;
  /** Eigenvectors as COLUMNS of a row-major (n, n) matrix, matching `values`. */
  vectors: Float64Array;
  /** Jacobi sweeps actually performed (diagnostic; convergence is typically 6-10). */
  sweeps: number;
}

/**
 * Cyclic Jacobi eigendecomposition of a symmetric (n, n) matrix.
 *
 * Each sweep annihilates every off-diagonal entry once with a Givens rotation chosen to
 * zero it; the off-diagonal Frobenius norm falls quadratically, so ~10 sweeps reach
 * double-precision accuracy. Eigenvalues come out on the diagonal and eigenvectors as
 * the accumulated rotation's columns. The input is not modified.
 */
export function jacobiEigen(matrix: ArrayLike<number>, n: number, maxSweeps = 100): Eigen {
  if (!Number.isInteger(n) || n < 1) throw invalidParam(`n must be a positive integer, got ${n}`);
  if (matrix.length !== n * n) throw invalidParam(`matrix has ${matrix.length} entries, expected ${n * n}`);

  const a = Float64Array.from(matrix);
  const v = new Float64Array(n * n);
  for (let i = 0; i < n; i++) v[i * n + i] = 1;

  // Scale-free stopping rule: stop once the off-diagonal mass is at the rounding floor.
  let frob = 0;
  for (let i = 0; i < a.length; i++) frob += a[i] * a[i];
  const floor = Math.sqrt(frob) * 1e-18;

  let sweeps = 0;
  for (; sweeps < maxSweeps; sweeps++) {
    let off = 0;
    for (let p = 0; p < n; p++) {
      for (let q = p + 1; q < n; q++) off += a[p * n + q] * a[p * n + q];
    }
    if (Math.sqrt(2 * off) <= floor) break;

    for (let p = 0; p < n - 1; p++) {
      for (let q = p + 1; q < n; q++) {
        const apq = a[p * n + q];
        if (apq === 0) continue;
        const app = a[p * n + p];
        const aqq = a[q * n + q];
        // t = tan(theta), the numerically stable root of t^2 + 2*theta*t - 1 = 0.
        const theta = (aqq - app) / (2 * apq);
        const t =
          theta >= 0
            ? 1 / (theta + Math.sqrt(theta * theta + 1))
            : -1 / (-theta + Math.sqrt(theta * theta + 1));
        const c = 1 / Math.sqrt(t * t + 1);
        const s = t * c;

        a[p * n + p] = app - t * apq;
        a[q * n + q] = aqq + t * apq;
        a[p * n + q] = 0;
        a[q * n + p] = 0;
        for (let k = 0; k < n; k++) {
          if (k === p || k === q) continue;
          const akp = a[k * n + p];
          const akq = a[k * n + q];
          const np = c * akp - s * akq;
          const nq = s * akp + c * akq;
          a[k * n + p] = np;
          a[p * n + k] = np;
          a[k * n + q] = nq;
          a[q * n + k] = nq;
        }
        for (let k = 0; k < n; k++) {
          const vkp = v[k * n + p];
          const vkq = v[k * n + q];
          v[k * n + p] = c * vkp - s * vkq;
          v[k * n + q] = s * vkp + c * vkq;
        }
      }
    }
  }

  // Sort descending, permuting the eigenvector columns with the eigenvalues.
  const order = Array.from({ length: n }, (_, i) => i).sort((i, j) => a[j * n + j] - a[i * n + i]);
  const values = new Float64Array(n);
  const vectors = new Float64Array(n * n);
  for (let j = 0; j < n; j++) {
    const src = order[j];
    values[j] = a[src * n + src];
    for (let i = 0; i < n; i++) vectors[i * n + j] = v[i * n + src];
  }
  canonicalizeSigns(vectors, n);
  return { values, vectors, sweeps };
}

/**
 * Fix each eigenvector's sign, in place: an eigenvector is only defined up to sign, so
 * two correct implementations can return MIRROR-IMAGE point clouds while every scalar
 * statistic agrees. The convention (architecture.md § Spectrum step 6, and the Python
 * side): if a column's largest-magnitude entry is negative, negate the whole column.
 * Ties on magnitude go to the first such entry by index.
 */
export function canonicalizeSigns(vectors: Float64Array, n: number): void {
  for (let j = 0; j < n; j++) {
    let best = 0;
    let bestAbs = -1;
    for (let i = 0; i < n; i++) {
      const m = Math.abs(vectors[i * n + j]);
      if (m > bestAbs) {
        bestAbs = m;
        best = i;
      }
    }
    if (vectors[best * n + j] < 0) {
      for (let i = 0; i < n; i++) vectors[i * n + j] = -vectors[i * n + j];
    }
  }
}

// --- the spectrum --------------------------------------------------------------------

/** Step 1: subtract each column's mean. Returns a fresh (rows, cols) matrix. */
export function centreColumns(A: ArrayLike<number>, rows: number, cols: number): Float64Array {
  if (A.length !== rows * cols) throw invalidParam(`matrix has ${A.length} entries, expected ${rows * cols}`);
  const out = new Float64Array(rows * cols);
  const mean = new Float64Array(cols);
  for (let r = 0; r < rows; r++) for (let c = 0; c < cols; c++) mean[c] += A[r * cols + c];
  for (let c = 0; c < cols; c++) mean[c] /= rows;
  for (let r = 0; r < rows; r++) for (let c = 0; c < cols; c++) out[r * cols + c] = A[r * cols + c] - mean[c];
  return out;
}

/** Step 2: G = Ac^T Ac, shape (cols, cols), accumulated symmetrically. */
export function gram(Ac: Float64Array, rows: number, cols: number): Float64Array {
  const G = new Float64Array(cols * cols);
  for (let i = 0; i < cols; i++) {
    for (let j = i; j < cols; j++) {
      let s = 0;
      for (let r = 0; r < rows; r++) s += Ac[r * cols + i] * Ac[r * cols + j];
      G[i * cols + j] = s;
      G[j * cols + i] = s;
    }
  }
  return G;
}

export interface SpectrumStats {
  /** lambda_i, descending, negatives clamped to 0 (they are float error). */
  eigenvalues: number[];
  /**
   * sigma_i = sqrt(lambda_i). The Gram route returns the null space as dust of order
   * sqrt(eps) * sigma_max (~1e-9 relative) rather than exact zeros; that is expected and
   * moves no statistic below, so nothing here asserts an exact zero.
   */
  singularValues: number[];
  totalVariance: number;
  /**
   * exp(-sum p ln p) — an ENTROPY of the spectrum, not a count of nonzero directions. It
   * equals `ceiling` only for a perfectly flat spectrum, and in practice sits well below
   * it even when the algebraic rank is saturated.
   */
  effectiveRank: number;
  stableRank: number;
  participationRatio: number;
  fracVarTop2: number;
  fracVarTop10: number;
  nDimsFor90pct: number;
  /** min(rows - 1, cols) — the ceiling on the ALGEBRAIC rank after centring. */
  ceiling: number;
}

/** Steps 3 and 4, given raw eigenvalues of the Gram matrix. */
export function spectrumStats(eigenvalues: ArrayLike<number>, rows: number, cols: number): SpectrumStats {
  const lam: number[] = [];
  for (let i = 0; i < eigenvalues.length; i++) lam.push(Math.max(eigenvalues[i], 0));
  lam.sort((a, b) => b - a);

  const total = lam.reduce((s, x) => s + x, 0);
  const ceiling = Math.min(rows - 1, cols);
  const zero: SpectrumStats = {
    eigenvalues: lam,
    singularValues: lam.map(() => 0),
    totalVariance: 0,
    effectiveRank: 0,
    stableRank: 0,
    participationRatio: 0,
    fracVarTop2: 0,
    fracVarTop10: 0,
    nDimsFor90pct: 0,
    ceiling,
  };
  if (!(total > 0)) return zero;

  let entropy = 0;
  let sumP2 = 0;
  for (const l of lam) {
    const p = l / total;
    if (p > 0) entropy -= p * Math.log(p);
    sumP2 += p * p;
  }
  const head = (k: number): number => lam.slice(0, k).reduce((s, x) => s + x, 0) / total;
  let cum = 0;
  let nDims = lam.length;
  for (let i = 0; i < lam.length; i++) {
    cum += lam[i];
    if (cum >= 0.9 * total) {
      nDims = i + 1;
      break;
    }
  }

  return {
    eigenvalues: lam,
    singularValues: lam.map(Math.sqrt),
    totalVariance: total,
    effectiveRank: Math.exp(entropy),
    stableRank: total / lam[0],
    participationRatio: 1 / sumP2,
    fracVarTop2: head(2),
    fracVarTop10: head(10),
    nDimsFor90pct: nDims,
    ceiling,
  };
}

export interface SpectrumResult extends SpectrumStats {
  /** PCA coordinates, row-major (rows, components). A PROJECTION, not the geometry. */
  coords: Float64Array;
  components: number;
  /** lambda_i / sum(lambda) for each retained component. */
  explainedVarianceRatio: number[];
  sweeps: number;
}

/** All six steps for one matrix. `rows` is |V| + specials; `cols` is d_model. */
export function spectrum(
  A: ArrayLike<number>,
  rows: number,
  cols: number,
  components = PCA_COMPONENTS,
): SpectrumResult {
  if (rows < 2) throw invalidParam(`a spectrum needs at least 2 rows, got ${rows}`);
  const Ac = centreColumns(A, rows, cols);
  const G = gram(Ac, rows, cols);
  const { values, vectors, sweeps } = jacobiEigen(G, cols);
  const stats = spectrumStats(values, rows, cols);

  const k = Math.min(components, cols);
  const coords = new Float64Array(rows * k);
  for (let r = 0; r < rows; r++) {
    for (let j = 0; j < k; j++) {
      let s = 0;
      for (let c = 0; c < cols; c++) s += Ac[r * cols + c] * vectors[c * cols + j];
      coords[r * k + j] = s;
    }
  }
  const explainedVarianceRatio =
    stats.totalVariance > 0 ? stats.eigenvalues.slice(0, k).map((l) => l / stats.totalVariance) : new Array(k).fill(0);

  return { ...stats, coords, components: k, explainedVarianceRatio, sweeps };
}

/**
 * The untrained random-init baseline the panel draws beside the trained spectrum
 * (FR-622): a matrix of the same shape whose rows are drawn from the embedding's own
 * initializer, N(0, 0.02^2). Any rank the trained model has that this one does not is
 * the part learning contributed.
 */
export function randomBaselineSpectrum(rows: number, cols: number, seed = 0, std = 0.02): SpectrumResult {
  const rand = sfc32(seed);
  const A = new Float64Array(rows * cols);
  for (let i = 0; i < A.length; i += 2) {
    const u1 = Math.max(rand(), 1e-12);
    const u2 = rand();
    const r = Math.sqrt(-2 * Math.log(u1)) * std;
    A[i] = r * Math.cos(2 * Math.PI * u2);
    if (i + 1 < A.length) A[i + 1] = r * Math.sin(2 * Math.PI * u2);
  }
  return spectrum(A, rows, cols);
}
